File: apps/api/app/schemas/admin_review.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import (
    AllowlistStatus,
    ContentStatus,
    ExtractedObjectType,
    LicenseStatus,
    ResourceSourceType,
    RobotsStatus,
    SourcePolicyDecision,
)
from app.schemas.extracted_duplicate import ExtractedDedupeResultRead


class ResourceProvenanceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source_type: ResourceSourceType
    url: str | None
    title: str | None
    author: str | None
    publisher: str | None
    license: str | None
    attribution: str | None
    summary: str | None
    quality_score: float | None
    domain_reputation_score: float | None = None
    content_length_score: float | None = None
    formula_density_score: float | None = None
    code_example_score: float | None = None
    educational_structure_score: float | None = None
    human_review_score: float | None = None
    status: ContentStatus


class SourcePolicyStatusRead(BaseModel):
    decision: SourcePolicyDecision
    allowlist_status: AllowlistStatus
    robots_status: RobotsStatus
    license_status: LicenseStatus
    attribution_required: bool
    attribution_present: bool
    reasons: list[str]
    host: str | None = None


class SourceQualityStatusRead(BaseModel):
    draft_quality_score: float | None = None
    resource_quality_score: float | None = None
    overall_score: float | None = None
    domain_reputation_score: float | None = None
    content_length_score: float | None = None
    formula_density_score: float | None = None
    code_example_score: float | None = None
    educational_structure_score: float | None = None
    human_review_score: float | None = None


class ExtractedObjectReviewRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    resource_id: int | None
    topic_job_id: int | None
    object_type: ExtractedObjectType
    payload_json: dict[str, Any]
    confidence_score: float | None
    quality_score: float | None
    duplicate_cluster_id: int | None
    status: ContentStatus
    extraction_method: str | None
    model_version: str | None
    created_at: datetime
    updated_at: datetime
    resource: ResourceProvenanceRead | None = None
    # Populated on detail GET for review UI (QP-046); list may leave these null.
    policy: SourcePolicyStatusRead | None = None
    quality: SourceQualityStatusRead | None = None
    duplicates: ExtractedDedupeResultRead | None = None


class ExtractedObjectEdit(BaseModel):
    payload_json: dict[str, Any] | None = None
    object_type: ExtractedObjectType | None = None
    confidence_score: float | None = Field(default=None, ge=0, le=1)
    quality_score: float | None = Field(default=None, ge=0, le=1)


class ReviewQueueResponse(BaseModel):
    items: list[ExtractedObjectReviewRead]

File: apps/web/lib/types/admin-review.ts
import type { ContentStatus, ResourceSourceType } from "@/lib/types/admin-import";

export type ExtractedObjectType = "concept" | "formula" | "example" | "question" | "flashcard";

export type DuplicateMatchType = "exact_raw" | "exact_normalized" | "near_normalized";

export type SourcePolicyDecision = "allow" | "review" | "deny";
export type AllowlistStatus = "allowed" | "denied" | "not_configured" | "not_applicable";
export type RobotsStatus = "allowed" | "disallowed" | "unknown" | "not_applicable";
export type LicenseStatus = "permissive" | "restrictive" | "unknown" | "missing";

export type ResourceProvenance = {
  id: number;
  source_type: ResourceSourceType;
  url: string | null;
  title: string | null;
  author: string | null;
  publisher: string | null;
  license: string | null;
  attribution: string | null;
  summary: string | null;
  quality_score: number | null;
  domain_reputation_score?: number | null;
  content_length_score?: number | null;
  formula_density_score?: number | null;
  code_example_score?: number | null;
  educational_structure_score?: number | null;
  human_review_score?: number | null;
  status: ContentStatus;
};

export type SourcePolicyStatus = {
  decision: SourcePolicyDecision;
  allowlist_status: AllowlistStatus;
  robots_status: RobotsStatus;
  license_status: LicenseStatus;
  attribution_required: boolean;
  attribution_present: boolean;
  reasons: string[];
  host: string | null;
};

export type SourceQualityStatus = {
  draft_quality_score: number | null;
  resource_quality_score: number | null;
  overall_score: number | null;
  domain_reputation_score: number | null;
  content_length_score: number | null;
  formula_density_score: number | null;
  code_example_score: number | null;
  educational_structure_score: number | null;
  human_review_score: number | null;
};

export type ExtractedDuplicateMatch = {
  source: string;
  object_id: number;
  title: string;
  match_type: DuplicateMatchType;
  similarity_score: number | null;
  status: ContentStatus | null;
  confidence_score: number | null;
};

export type CanonicalSuggestion = {
  kind: string;
  object_id: number;
  title: string;
  reason: string;
};

export type ExtractedDedupeResult = {
  extracted_object_id: number;
  object_type: ExtractedObjectType;
  raw_text_hash: string;
  normalized_text_hash: string;
  normalized_text: string;
  matches: ExtractedDuplicateMatch[];
  suggested_canonical: CanonicalSuggestion;
};

export type ReviewQueueItem = {
  id: number;
  resource_id: number | null;
  topic_job_id: number | null;
  object_type: ExtractedObjectType;
  payload_json: Record<string, unknown>;
  confidence_score: number | null;
  quality_score: number | null;
  duplicate_cluster_id: number | null;
  status: ContentStatus;
  extraction_method: string | null;
  model_version: string | null;
  created_at: string;
  updated_at: string;
  resource: ResourceProvenance | null;
  policy?: SourcePolicyStatus | null;
  quality?: SourceQualityStatus | null;
  duplicates?: ExtractedDedupeResult | null;
};

export type ReviewQueueResponse = {
  items: ReviewQueueItem[];
};

export type ReviewQueueEditInput = {
  payloadJson?: Record<string, unknown>;
  objectType?: Extract<ExtractedObjectType, "question" | "flashcard">;
  qualityScore?: number | null;
  confidenceScore?: number | null;
};

export type ReviewQueueStatus = "draft" | "approved" | "rejected";

export type PublishedObjectRef = {
  kind: "concept" | "question" | "flashcard";
  id: number;
};

export type PublishDuplicateMatch = {
  question_id: number;
  title: string;
  match_type: string;
  similarity_score: number | null;
};

export type PublishReviewResult = {
  extracted_object_id: number;
  object_type: ExtractedObjectType;
  published: PublishedObjectRef;
  duplicate_warnings: PublishDuplicateMatch[];
};
